custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml

```diff
@@ -293,10 +308,38 @@ secondary_entities:
     dps:
       - id: 129
         name: button
+        optional: true
         type: boolean
-
-
-
-
-
-
+  # add additional binary_sensor for each fault code
+  # add fault codes to mapping
+  - entity: sensor
```

Comment:
I don't think this catch-all fault sensor is needed, since it will show up under the error attribute of the vacuum.  Also you can add a mapping to that to translate states into something more readable, rather than having binary sensors for each error, as there are typically dozens of possible errors for vacuums.

In any case, this sensor can only have a class of enum if there is a mapping to a list of possible values it can have (and that list must be  complete, or HA starts throwing errors when it gets unexpected values).

Reply:
Ah making this an enum was accidental, thanks for picking that up.

I want this to report the name of the fault, or just the number if there wasn't a mapping - e.g. 'Trapped' (for a known fault code) or e.g. '12345' (for an unknown fault code). This allows an error state to surface, and further details of the error to be visible.

Regardless of the `type`, mappings for the `name: error` sensor are more-or-less ignored - an error state is surfaced but the details of the fault code are obscured. My thinking was to then have a 'fault code' sensor, which would surface known errors by name, and unknown errors by number. This didn't seem to work - if a mapping is specified, HASS wants it to be an enum and throws an error if it's not (not sure why, or if i'm doing something wrong - see error below). If i make it an enum, it throws an error if an unknown fault code is encountered (which makes sense). My thought from there is to create a single sensor to expose the error code, and separate binary sensors (as hinted in the docs RE bitfield) for each error code.

Without `class: enum` specified, any `mapping` produces the following error:
```hass           | 2023-06-13 10:14:15.488 DEBUG (MainThread) [custom_components.tuya_local.device] Retrying after exception <class 'ValueError'> Sensor sensor.kogan_lx10_robot_vacuum_fault_code is providing enum options, but is missing the enum device class (2/3)```

Reply:
You can see the fault code on the error attribute of the vacuum.  A mapping there won't get caught up in HA's overly strict policing of sensor values.

Reply:
Thanks, so I can get the error via `state_attr('vacuum.kogan_lx10_robot_vacuum', 'error')` but as far as I can tell it's not otherwise surfaced to the user. I don't particularly care about the logbook, I care about being able to see _why_ there's an error, without having to go find the vacuum - this information is surfaced in the Tuya app & IMO it should be visible in HASS without additional work required by the end-user.

Reply:
You can also surface it in most general cards (not the Glance card IIRC).  The Entity card has a UI for selecting attributes, with the Entities card, I think you still have to edit the yaml.  Or you can see it if you open the "Attribute" drawer in the more info window for the entity.

Reply:
🤦 that's exactly what I wanted, thanks